contracts/root/StakeManager.sol

```diff
@@ -6,182 +6,244 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
 
 import { ECVerify } from "../lib/ECVerify.sol";
 import { BytesLib } from "../lib/BytesLib.sol";
+import { PriorityQueue } from "../lib/PriorityQueue.sol";
+
 import { Lockable } from "../mixin/Lockable.sol";
 import { RootChainable } from "../mixin/RootChainable.sol";
 
 import { StakeManagerInterface } from "./StakeManagerInterface.sol";
 import { RootChain } from "./RootChain.sol";
+import { ValidatorSet } from "./ValidatorSet.sol"; 
 
 
 contract StakeManager is StakeManagerInterface, RootChainable, Lockable {
   using SafeMath for uint256;
+  using SafeMath for uint8;
   using ECVerify for bytes32;
 
+  PriorityQueue stakeQueue;
+  ValidatorSet validatorSet;
   // token object
   ERC20 public tokenObj;
 
-  // validator threshold
-  uint256 public _validatorThreshold = 0;
-
-  // pool of stakers and validators
-  struct Stake {
-    uint256 amount;
-    address staker;
-  }
-
-  // map of stakers (address => staked index)
-  mapping(address => uint256) public stakers;
+  // The randomness seed of the epoch.
+  // This is used to determine the proposer and the validator pool
+  bytes32 public epochSeed = keccak256(abi.encodePacked(block.difficulty + block.number + now));
 
-  // stake list
-  Stake[] public stakeList;
+  event ThresholdChange(uint256 newThreshold, uint256 oldThreshold);
+  //optional event to ack unstaking
+  event UnstakeInit(address indexed user, uint256 amount, uint256 total, bytes data); 
 
+  // validator threshold
+  uint256 public _validatorThreshold = 0;
   // total stake
-  uint256 public totalStake;
+  uint256 public totalStake = 0;
+  // current epoch
+  uint256 public currentEpoch = 0;
 
-  // The randomness seed of the epoch.
-  // This is used to determine the proposer and the validator pool
-  bytes32 public epochSeed = keccak256(abi.encodePacked(block.difficulty + block.number + now));
+  //Todo: dynamically update
+  uint256 public minStakeAmount = 0;
+  uint256 public minLockInPeriod = 1; //(unit epochs)
+  uint256 public stakingIdCount = 0;  // just a counter/index to map it with PQ w/address
 
-  //
-  // Events
-  //
+  enum State { isStaker, isValidator, isProposer, isBaned, isUnstaking }
```

Comment:
Change to `ValidatorStatus`. Status can be any one of these - `WAITING`, `SET`, & `LEAVING`, `LEFT` or something similar.